department-of-veterans-affairs/vets-website
# [department-of-veterans-affairs/va-virtual-agent#1105] Decision Letters

## Summary
- Add `cardActionMiddleware` & `activityMiddleware` to WebChat rendering if the feature toggle `virtualAgentDecisionLetterDownloadTracking` is on; otherwise, render using original behavior
  - The new `cardActionMiddleware` will post to Google Analytics by calling the `recordEvent` function if the event is determined to be a decision letter download
- Add tests for new `cardActionMiddleware` function
  - Verify that `recordEvent` is called once with expected parameters and `next` is called once when it's a decision letter downloaded
  - Verify that `recordEvent` is not called and `next` is called when it's not a decision letter

Team: Virtual Agent Chatbot
Maintenance Owner: Yes
Flipper End Date: 11/1/23


## Related issue(s)
- `va-virtual-agent` issue: https://github.com/department-of-veterans-affairs/va-virtual-agent/issues/1105
- `vets-website` feature flag: https://github.com/department-of-veterans-affairs/vets-website/pull/25855
- `vets-api` feature flag: https://github.com/department-of-veterans-affairs/vets-api/pull/13943


## Testing done
- Before this change, when a user clicked to download a decision letter from within the Chatbot, there would be no recorded event in Google analytics. With this change, there will be an event recorded.
- To verify this change:
  - Add a console.log message to the new `cardActionMiddleware` function in `Webchat.jsx`. This is needed because the are no visual changes to this this modification and Google Analytics does not track data from local runs.
  - Run `vets-website` and `vets-api` locally.
  - Enable the flipper feature toggle `virtualAgentDecisionLetterDownloadTracking`.
  - Go to http://localhost:3001/contact-us/virtual-agent/ and open your dev console.
  - Click **Start Chat** in the Chatbot and send a message saying "decision letter"
  - The bot will ask if you want to login. Say yes and follow the steps.
  - Once the bot returns and gives you the decision letters, click one of the letters.
  - Look for your message in the console.
  - Click another decision letter and verify that each time you click to download, there is a message.
  - Repeat these steps with the feature flag off and verify that no message is shown.
Test completed:
  - We ran this locally using the steps mentioned above and verified that the message showed up only when the feature flag was on and that the message showed up each time the button was clicked.
  - We created unit tests for the new function to make sure the event recorded only when it was a decision letter download.


## What areas of the site does it impact?
Only the Chatbot located at /contact-us/virtual-agent/.

## Acceptance criteria

### Quality Assurance & Testing


### Error Handling


### Authentication


### :warning: Team Sites (only applies to modifications made to the VA.gov header) :warning:


## Requested Feedback

(OPTIONAL) _What should the reviewers know in addition to the above. Is there anything specific you wish the reviewer to assist with. Do you have any concerns with this PR, why?_
We were wondering about the `recordEvent` arguments.
- Is there a standard way to determine what type of `event` to put. In this case, we are tracking every time a user downloads a decision letter through our bot. Right now, we have the event as a `cta-button-click` but technically, this isn't a CTA as we're not asking the user to do anything. Is there an event that says something that makes more sense for this type of event (button click or file download)?
- The label looks like it's normally exactly what's written on the button. In this case, we are looking to track all button clicks total but these button labels are dynamic so we've generalized it to `decisionLetter`. Is that okay?
